fem/integ/bilininteg_hdiv_ea.cpp

```diff
@@ -0,0 +1,336 @@
+
+         const real_t (&Bi1)[MQ1][MD1] = (ic == 0) ? r_Bc : r_Bo;
+         const real_t (&Bi2)[MQ1][MD1] = (ic == 0) ? r_Bo : r_Bc;
+
+         for (int idx_j = 0; idx_j < NDOF; ++idx_j)
+         {
+            const int jc = idx_j / (D1D*(D1D-1));
+            const int idx_jj = idx_j % (D1D * (D1D-1));
+            const int jx = (jc == 0) ? idx_jj%D1D : idx_jj%(D1D-1);
+            const int jy = (jc == 0) ? idx_jj/D1D : idx_jj/(D1D-1);
+
+            const real_t (&Bj1)[MQ1][MD1] = (jc == 0) ? r_Bc : r_Bo;
+            const real_t (&Bj2)[MQ1][MD1] = (jc == 0) ? r_Bo : r_Bc;
+
+            real_t val = 0.0;
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               for (int qy = 0; qy < Q1D; ++qy)
+               {
+                  const double coeff = s_D[ic + jc*2][qx][qy];
+                  val += coeff*Bi1[qx][ix]*Bi2[qy][iy]*Bj1[qx][jx]*Bj2[qy][jy];
+               }
+            }
+            if (add)
+            {
+               M(idx_i, idx_j, e) += val;
+            }
+            else
+            {
+               M(idx_i, idx_j, e) = val;
+            }
+         }
+      }
+   });
+}
+
+// For H(div) mass, Bo and Bc are the basis evaluation operators, and the
+// pa_data corresponds to a (potentially symmetric) matrix coefficient.
+// coeff_dim must be 6 or 9 depending on symmetry.
+//
+// For div-div, Bc is the derivative evaluation operator, and pa_data
+// corresponds to a scalar coefficient. coeff_dim must be 1.
+//
+// These two integrators are distinguished using coeff_dim.
+template<int T_D1D = 0, int T_Q1D = 0>
+static void EAHdivAssemble3D(const int NE,
+                             const Array<real_t> &Bo_,
+                             const Array<real_t> &Bc_,
+                             const int coeff_dim,
+                             const Vector &pa_data,
+                             Vector &ea_data,
+                             const bool add,
+                             const int d1d = 0,
+                             const int q1d = 0)
+{
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+   MFEM_VERIFY(D1D <= DeviceDofQuadLimits::Get().HDIV_MAX_D1D, "");
+   MFEM_VERIFY(Q1D <= DeviceDofQuadLimits::Get().HDIV_MAX_Q1D, "");
+   const int NDOF_C = (D1D-1)*(D1D-1)*D1D;
+   const int NDOF = 3*NDOF_C;
+   const auto Bo = Reshape(Bo_.Read(), Q1D, D1D-1);
+   const auto Bc = Reshape(Bc_.Read(), Q1D, D1D);
+   const auto D = Reshape(pa_data.Read(), Q1D, Q1D, Q1D, coeff_dim, NE);
+   const bool symmetric = (coeff_dim == 6);
+   auto M = Reshape(add ? ea_data.ReadWrite() : ea_data.Write(), NDOF, NDOF, NE);
+   mfem::forall_2D(NE, NDOF, 1, [=] MFEM_HOST_DEVICE (int e)
+   {
+      constexpr int MD1 = T_D1D ? T_D1D : DofQuadLimits::HDIV_MAX_D1D;
+      constexpr int MQ1 = T_Q1D ? T_Q1D : DofQuadLimits::HDIV_MAX_Q1D;
+      // Load Bo and Bc matrices into registers
+      real_t r_Bo[MQ1][MD1];
+      real_t r_Bc[MQ1][MD1];
```

Comment:
any worry that at higher order this is going to spill?

Reply:
This is the usual approach for MFEM's EA kernels (diffusion, mass, etc.). We could potentially be more efficient at high orders by better using shared memory, but I haven't noticed that this has been a bottleneck in my testing.